include/boost/geometry/algorithms/densify.hpp

```diff
@@ -0,0 +1,368 @@
+// Boost.Geometry
+
+// Copyright (c) 2017-2018, Oracle and/or its affiliates.
+
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_DENSIFY_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_DENSIFY_HPP
+
+
+#include <boost/geometry/algorithms/clear.hpp>
+#include <boost/geometry/algorithms/detail/convert_point_to_point.hpp>
+#include <boost/geometry/algorithms/not_implemented.hpp>
+#include <boost/geometry/core/closure.hpp>
+#include <boost/geometry/core/point_type.hpp>
+#include <boost/geometry/core/tag.hpp>
+#include <boost/geometry/core/tags.hpp>
+#include <boost/geometry/strategies/default_strategy.hpp>
+#include <boost/geometry/strategies/densify.hpp>
+#include <boost/geometry/util/condition.hpp>
+#include <boost/geometry/util/range.hpp>
+
+#include <boost/range/size.hpp>
+#include <boost/range/value_type.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace densify
+{
+
+template <typename Range>
+struct push_back_policy
+{
+    typedef typename boost::range_value<Range>::type point_type;
+
+    inline explicit push_back_policy(Range & rng)
+        : m_rng(rng)
+    {}
+
+    inline void apply(point_type const& p)
+    {
+        range::push_back(m_rng, p);
+    }
+
+private:
+    Range & m_rng;
+};
+
+template <typename Range, typename Point>
+inline void convert_and_push_back(Range & range, Point const& p)
+{
+    typename boost::range_value<Range>::type p2;
+    geometry::detail::conversion::convert_point_to_point(p, p2);
+    range::push_back(range, p2);
+}
+
+template <bool AppendLastPoint = true>
+struct densify_range
+{
+    template <typename Geometry, typename GeometryOut, typename T, typename Strategy>
+    static inline void apply(Geometry const& rng, GeometryOut & rng_out,
+                             T const& len, Strategy const& strategy)
+    {
+        typedef typename boost::range_value<Geometry>::type point_t;
+
+        std::size_t count = boost::size(rng);
+
+        if (count == 0)
+            return;
+        
+        push_back_policy<GeometryOut> policy(rng_out);
+
+        for (std::size_t i = 1 ; i < count ; ++i)
+        {
+            point_t const& p0 = range::at(rng, i - 1);
+            point_t const& p1 = range::at(rng, i);
+
+            convert_and_push_back(rng_out, p0);
+
+            strategy.apply(p0, p1, policy, len);
```

Comment:
Any calculation related to lengths is done in strategies. The strategy only generates intermediate points which means that it may generate no points and that the responsibility to return segment's endpoints is not the algorithm's side.

AppendLastPoint is here to distinguish between open and closed geometries. And yes, if the strategy doesn't generate any points then the condition below appends the last endpoint.